### PR TITLE
feat(autoprofile): rework prebuilds

### PR DESCRIPTION
**Pull request by kirrg001**
refs 149946

Move away from manual prebuilds for the autoprofile package to using [prebuildify](https://github.com/prebuild/prebuildify).

The PR **replaces the prebuilds** for darwin x64 and linux x64 (musl + glibc) **and adds more prebuilds for** darwin arm64 (m1), linux arm64 and armv6 and armv7 (raspberry pi).

This is the first iteration. The next iteration is to get rid of the manual prebuilds in the shared metrics package.

# Testing / QA

Installing the autoprofiler on darwin x64 **from local source without prebuild**
> npm i ~/dev/instana/nodejs/packages/autoprofile/instana-autoprofile-3.1.0.tgz --save --verbose
> added 39 packages, and audited 40 packages in 14s

Installing the autoprofiler on darwin x64 **from local source with prebuild**
> npm i ~/dev/instana/nodejs/packages/autoprofile/instana-autoprofile-3.1.0.tgz --save --verbose
> added 39 packages, and audited 40 packages in 3s

Installing the autoprofiler on darwin x64 **from npm including manual prebuilds**
> npm i @instana/autoprofile@latest --save
> added 38 packages, and audited 39 packages in 13s

-> **caused by a bug in the node-gyp-fallback script.**

Installing the autoprofiler on linux x64 musl **from npm including manual prebuilds**
>  => [3/3] RUN npm install @instana/autoprofile --save --verbose                                                                  10.8s

Installing the autoprofiler on linux x64 musl **from local source with prebuild**
> => [4/4] RUN npm install instana-autoprofile-3.1.0.tgz --save --verbose                                                          6.3s

Installing the autoprofiler on linux x64 musl **from local source without prebuild**
>  => [5/5] RUN npm install instana-autoprofile-3.1.0.tgz --save --verbose                                                         18.3s

Installing the autoprofiler on linux x64 glibc **from npm including manual prebuilds**
> => [4/4] RUN npm install @instana/autoprofile --save --verbose                                                                   7.8s
 
Installing the autoprofiler on linux x64 glibc **from local source with prebuild**
> => [4/4] RUN npm install instana-autoprofile-3.1.0.tgz --save --verbose                                                          6.5s
 
Installing the autoprofiler on linux x64 glibc **from local source without prebuild**
> => [4/4] RUN npm install instana-autoprofile-3.1.0.tgz --save --verbose                                                         16.6s

Installing the autoprofiler on linux arm64 glibc **from npm including manual prebuilds**
> => [4/4] RUN npm install @instana/autoprofile --save --verbose                                                                 118.7s
  
Installing the autoprofiler on linux arm64 glibc **from local source with prebuild**
> => [4/4] RUN npm install instana-autoprofile-local-with-prebuild.tgz --save --verbose                                 31.1s.

# Tasks

- [x] create a pull request for https://github.com/prebuild/prebuildify-cross/issues/18
- [x] bump prebuildify-cross to 5.1.0
- [x] manually test on darwin
- [x] manually test on linux
- [x] code QA

### CODE REVIEW

**Comment by kirrg001** [comment]
RFR but do not merge. We need to wait for the merge & release of https://github.com/prebuild/prebuildify-cross/issues/18.

**Comment by aryamohanan** [comment]
Awesome work 🥳

File: packages/autoprofile/CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+## Prebuilds
```
Comment (kirrg001) [inline packages/autoprofile/CONTRIBUTING.md:1]: Re-add the docs lines:

> From https://github.com/lando/lando/issues/3533
> $ sudo ln -s ~/.rd/docker.sock /var/run/docker.sock

File: package.json [hunk 2]
```diff
@@ -166,7 +166,9 @@
     "mysql2": "^3.2.3",
     "nats": "^2.11.0",
     "nats-v1": "npm:nats@^1.4.12",
+    "node-abi": "^3.54.0",
     "node-fetch": "^2.6.1",
+    "node-gyp": "^10.0.1",
```
Comment (aryamohanan) [inline package.json:171]: Qs: why we installed node-gyp? For  `prebuildify` node-gyp-build is required right?

Comment (kirrg001) [reply package.json:171]: Yeah
https://github.com/prebuild/prebuildify-cross?tab=readme-ov-file#install

Node-gyp is shipped with npm anyway, but I really didn't like that we rely on the npm version.

File: packages/autoprofile/CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    - linux/arm64
+    - linux/armv6
+    - linux/armv7
+
+```sh
+node scripts/prebuilds.js                               [build all abi version for darwin and linux]
+node scripts/prebuilds.js --os=darwin                   [build all abi version only for darwin]
+node scripts/prebuilds.js --os=linux                    [build all abi version only for linux]
+node scripts/prebuilds.js --os=linux --arch linux-arm64 [build all abi version only for linux arm64]
+node scripts/prebuilds.js --abi=14.0.0                  [build specific abi version]
```
[the remark this answers is not in the document]
Comment (kirrg001) [reply packages/autoprofile/CONTRIBUTING.md:19]: Ah good catch. I will update the section

> ## Adding support for a new Node.js version

File: packages/autoprofile/scripts/prebuilds.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+const abi = require('node-abi');
+const yargs = require('yargs/yargs');
+const { hideBin } = require('yargs/helpers');
+const argv = yargs(hideBin(process.argv)).argv;
+
+const semver = require('semver');
+const childProcess = require('child_process');
+
+const minNodeVersion = require('../package.json').engines.node;
+const excludeTargets = ['15.0.0', '17.0.0', '19.0.0'];
```
[the remark this answers is not in the document]
Comment (kirrg001) [reply packages/autoprofile/scripts/prebuilds.js:24]: Yeah hmmm I am wondering if we should remove the whole logic if no abi version is present.
It was helpful for the initial build, but in the future we will only build a specific abi version.

File: package.json [hunk 5]
```diff
@@ -175,6 +177,8 @@
     "pino-v6": "npm:pino@^6.13.3",
     "pino-v7": "npm:pino@^7.8.1",
     "pkg-up": "^3.1.0",
+    "prebuildify": "^5.0.1",
```
Comment (aryamohanan) [inline package.json:180]: Suggestion: `prebuildify` version 6.0.0  released, can we use the latest? https://www.npmjs.com/package/prebuildify/v/6.0.0

Comment (kirrg001) [reply package.json:180]: Ah, they released a new major version during my work. I don't prefer to jump to a X.0.0 release. And the changes do not look interesting to me.

**Comment by kirrg001** [comment]
prebuildify-cross released 5.1.0. Will update 👍

**Comment by kirrg001** [comment]
Closing because of wrong branch name.
https://github.com/instana/nodejs/pull/1135

**Pull request closed without merging**